[marquestye] iP

marquestye opened the pull request:
# Duck
> [!IMPORTANT]
> “Quack.” – :duck: ([source](https://www.youtube.com/watch?v=aqCxlxclyzo))

Duck frees your mind of having to remember things you need to do. It's,  
- text-based
- easy to learn 
- ~~FAST~~ _SUPER FAST_ to use 

All you need to do is,  
1. download it from [here](https://nus-cs2103-ay2324s1.github.io/website/admin/ip-w4.html).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😉

And it is **FREE**!

## Features:
- [x] Managing tasks
- [ ]  Managing deadlines (coming soon)
- [ ]  Reminders (coming soon)
---
If you're a Java programmer, you can use it to practice Java too. Here's the main method:
```
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

marquestye committed:
renamed to Duck, replaced default greeting

marquestye committed:
added echoing of input

marquestye committed:
added simple task list

marquestye committed:
added marking and unmarking

marquestye committed:
added different types of tasks

marquestye committed:
added text ui testing

marquestye committed:
basic error handling

marquestye committed:
added task deletion (inefficient)

marquestye committed:
fixed list bug

marquestye committed:
Refactor to OOP with exceptions

marquestye committed:
fix and update new tests

marquestye committed:
Update README

marquestye committed:
Add basic saving function

marquestye committed:
Add history editing for marking

marquestye committed:
Add history editing for deletion

marquestye committed:
Merge branch 'branch-Level-7'

marquestye committed:
Add datetime recognition to DeadlineTask

marquestye committed:
Merge branch 'branch-Level-8'

marquestye committed:
Refactor into packages with OOP

marquestye committed:
Merge branch 'branch-A-MoreOOP'

marquestye committed:
Merge remote-tracking branch 'origin/add-gradle-support'

marquestye committed:
Fix delete not reflecting in storage

marquestye committed:
Modify build.gradle for build and run

marquestye committed:
Add basic JUnit tests

marquestye committed:
Merge branch 'branch-A-JUnit'

marquestye committed:
Add JavaDocs for command package

marquestye committed:
Add JavaDocs for task package

marquestye committed:
Add JavaDocs for task package

marquestye committed:
Fix styling for command package

marquestye committed:
Fix styling for duck package

marquestye committed:
Fix styling for task package

marquestye committed:
Configure gradle checkstyle

marquestye committed:
Fix styling raised by gradle

marquestye committed:
Add find command

marquestye committed:
Merge branch 'branch-A-JavaDoc'

marquestye committed:
Resolve various merge conflicts

marquestye committed:
Resolve merge conflicts

marquestye committed:
Fix unrecognised find command bug

marquestye committed:
Add javadocs for methods related to find

ngeeyonglim commented on src/main/java/duck/Duck.java:
```diff
@@ -0,0 +1,59 @@
+package duck;
+
+import java.util.ArrayList;
+
+import duck.command.Command;
+
+import duck.task.Task;
+import duck.task.TaskList;
```
Good ordering of imports into different categories

ngeeyonglim commented on src/main/java/duck/Duck.java:
```diff
@@ -0,0 +1,59 @@
+    private Storage storage;
+    private TaskList tasks;
+
+    /**
+     * Initialises the Duck chatbot and attempts to load tasks from storage.
+     * 
+     * @param filePath The path to the file where the tasks are stored.
+     */
```
Good use of javadocs

ngeeyonglim commented on src/main/java/duck/Duck.java:
```diff
@@ -0,0 +1,59 @@
+     */
+    public void run() {
+        ui.showWelcomeMessage();
+        boolean isExit = false;
```
Good variable name

ngeeyonglim commented on src/main/java/duck/Parser.java:
```diff
@@ -0,0 +1,139 @@
+    public static Command parse(String input) throws DuckException {
+        String[] splitInput = input.split(" ", 2);
+        String command = splitInput[0].toUpperCase();
+        switch (command) {
+        case "LIST":
```
Good indentation practice

ngeeyonglim commented on src/main/java/duck/Parser.java:
```diff
@@ -0,0 +1,139 @@
+        }
+    }
+
+    private static Task parseTodo(String dataString) throws DuckException {
```
Perhaps consider changing the method name to show that a Task object will be returned?

ngeeyonglim commented on src/main/java/duck/Storage.java:
```diff
@@ -0,0 +1,122 @@
+import java.io.FileWriter;
+import java.io.IOException;
+
+import java.util.ArrayList;
+import java.util.Scanner;
```
You may consider splitting these 2 imports up as they have slightly different functionalities?

ngeeyonglim commented on src/main/java/duck/Storage.java:
```diff
@@ -0,0 +1,122 @@
+        File dataDir = new File(filePath).getParentFile();
+        dataDir.mkdir(); // Create directory only if it doesnt already exist
+
+        File taskFile;
```
Good use of splitting code blocks with different functions

ngeeyonglim commented on src/main/java/duck/command/AddCommand.java:
```diff
@@ -0,0 +1,31 @@
+ * Represents an executable command which adds a task.
+ */
+public class AddCommand extends Command {
+    private Task newTask;
```
Good variable name

ngeeyonglim reviewed (commented):
LGTM. Code is very clean and follows coding standards.

ryantzr1 commented on src/test/java/duck/task/TodoTaskTest.java:
```diff
@@ -0,0 +1,38 @@
+
+    @Test
+    public void toString_success() {
+        TodoTask task = new TodoTask("do laundry", false);
```
I like how you have multiple test cases here. You can consider to add in negative cases as well?

ryantzr1 commented on src/test/java/duck/task/TodoTaskTest.java:
```diff
@@ -0,0 +1,38 @@
+    public void stringify_success() {
+        TodoTask task = new TodoTask("do laundry", false);
+        assertEquals("T010/do laundry", task.stringify());
+
```
Perhaps you could set up a constant variable when naming "do laundry, finish hw etc. as these may be considered to be magic strings.

ryantzr1 commented on src/test/java/duck/task/TaskListTest.java:
```diff
@@ -0,0 +1,65 @@
+        TaskList taskList = new TaskList();
+        try {
+            taskList.mark(1);
+            fail();
```
Perhaps you were trying to use "unmark" instead of "mark" here?

ryantzr1 commented on src/test/java/duck/ParserTest.java:
```diff
@@ -0,0 +1,70 @@
+public class ParserTest {
+    @Test
+    public void parse_unknownCommand_exceptionThrown() {
+        try {
```
Good use of variable names

ryantzr1 commented on src/main/java/duck/task/TodoTask.java:
```diff
@@ -0,0 +1,48 @@
+
+    @Override
+    public String stringify() {
+        return "T" + super.stringify();
```
Perhaps you could change the naming of this function because having a stringify and a toString function together may seem a little confusing.

ryantzr1 commented on src/main/java/duck/task/TaskList.java:
```diff
@@ -0,0 +1,143 @@
+            tasks.remove(index);
+        } catch (IndexOutOfBoundsException e) {
+            throw new DuckException("Error - invalid task number.");
+        }
```
Perhaps you could use a constant variable for the error message because it repeats multiple times.

ryantzr1 commented on src/main/java/duck/task/Task.java:
```diff
@@ -0,0 +1,97 @@
+     */
+    public void mark() throws DuckException {
+        if (this.isDone == true) {
+            throw new DuckException("Error - Task is already marked.");
```
perhaps you can simplify this into `if (this.isDone)` directly

ryantzr1 commented on src/main/java/duck/task/EventTask.java:
```diff
@@ -0,0 +1,91 @@
+        if (obj instanceof EventTask) {
+            EventTask other = (EventTask) obj;
+            return super.equals(other) && start.equals(other.start) && end.equals(other.end);
+        }
```
Perhaps you can split the boolean checking up in order to reduce complicated expressions

ryantzr1 commented on src/main/java/duck/Ui.java:
```diff
@@ -0,0 +1,106 @@
+    }
+
+    public void showAllTasks(TaskList tasks) {
+        if (tasks.getTaskCount() == 0) {
```
good use of variable naming!

ryantzr1 reviewed (commented):
LGTM! Just a few small points to take note!

marquestye committed:
Add javafx

marquestye committed:
Merge branch 'branch-Level-10'

marquestye committed:
Delete unneeded DialogBox file

marquestye committed:
Add assert statements

marquestye committed:
Improve task deletion logic for storage

Task deletion repeats some lines of code.

The code duplication results in messy logic that confuses readers.

Let's reorder the code.

Doing so allows for a better flow of logic, improving code clarity.

marquestye committed:
Improve storage load task readability

Task loading uses arrowhead style code.

Deep nesting makes it harder for readers to keep track of logic.

Let's utilise guard clauses to reduce nesting of the happy path.

Doing so results in greater readability.

marquestye committed:
Merge pull request #3 from marquestye/branch-A-CodeQuality

Branch-A-CodeQuality

marquestye committed:
Merge branch 'master' into branch-A-Assertions

marquestye committed:
Merge pull request #2 from marquestye/branch-A-Assertions

Branch-A-Assertions

marquestye committed:
Add tagging of tasks

marquestye committed:
Merge branch 'branch-C-Tagging'

marquestye committed:
Merge branch 'master' of https://github.com/marquestye/ip

marquestye committed:
Add screenshot and UG

marquestye committed:
Correct typos and wording in UG

marquestye committed:
Fix UG formatting errors

marquestye committed:
Fix more UG formatting

marquestye committed:
Update tests
